### PR TITLE
[WIP] Docs review: → arrow awesomeness

### PR DESCRIPTION
**Pull request by felipenoris**

### CODE REVIEW

**Comment by martinholters** [comment]
In light of #16432, we might want to go with the return type syntax from there for cases where the docstring states the return type in the header. Maybe a general form like
```Julia
foo(v::Vector{T})::(Int, T) → (foo_count, foo_factor)
```
where return type and return value names may be omitted if they are more cluttering than helpful could be adopted?

**Comment by nalimilan** [comment]
@martinholters Good idea, but the actual syntax is
```julia
foo(v::Vector{T})::Tuple{Int, T}
```

Adding the names after the arrow sounds like a good strategy.

**Comment by martinholters** [comment]
@nalimilan Of course, thanks.

**Comment by felipenoris** [comment]
Great! I'll do it.

**Comment by felipenoris** [comment]
@nalimilan , @martinholters : maybe it would be easier to use the actual language syntax. It's easy for me to replace all `→` with `::` , even if the docstring won't match the exact return types declared on methods.

**Comment by nalimilan** [comment]
@felipenoris I think it's fine if they don't match the return type *declarations* as long as they match the *actual* return type. In general it should be the case, or there's a bug in the current documentation.

**Comment by felipenoris** [comment]
Docstrings usually tell about return types only. I propose I change this:
```julia
    daysofweekinmonth(dt::TimeType) → Int
```

to this:
```julia
    daysofweekinmonth(dt::TimeType) :: Int
```

The default use case would be just to copy-paste method declaration into docstrings when the return type is declared.

For @martinholters example:
```julia
    foo(v::Vector{T})::(Int, T) → (foo_count, foo_factor)
```

that docstring shouldn't exist. It would be written as:
```julia
    foo(v::Vector{T}) :: (Int, T)
```

My point is that there's no need to name variables for return values. We usually just explain what they are in the documentation.

**Comment by felipenoris** [comment]
Most methods return just one value, and are documented as returning types, as in:
```julia
    isinteger(x) → Bool
```

For tuples, I see both cases on current docstrings. This one uses types:
```julia
    splitdir(path::AbstractString) → (AbstractString, AbstractString)
```

This one give names to return values.
```julia
    recvfrom(socket::UDPSocket) → (address, data)
```

I think there's no correct answer for this. But my guess is that using `::` with return types is less verbose and easier to use in most cases.

**Comment by omus** [comment]
If we're sticking to actual Julia syntax then it should be:
```julia
splitdir(path::AbstractString)::Tuple{AbstractString, AbstractString}
```
I personally think that we should probably keep using the arrow syntax for documentation as the documentation signature isn't actually Julia code but is rather pseudo code:

```julia
isinteger(x) -> Bool
splitdir(path::AbstractString) -> (AbstractString, AbstractString)
recvfrom(socket::UDPSocket) -> (address, data)
mkdir(path[, mode])
```

**Comment by felipenoris** [comment]
@omus, I agree, which means that this PR is done in terms of coding. So it´s up to admins to accept or not the change from `->` to `→`.

**Comment by nalimilan** [comment]
@omus I don't see the point of using a different syntax when we could use the real one. This doesn't make anything clearer.

@felipenoris Giving names to return values can be useful when returning a tuple. Though it should only be used when really needed, and not considered a requirement. Anyway, we should absolutely avoid writing `::(name1, name2)`, which looks like Julia syntax but isn't and creates confusion.

**Comment by omus** [comment]
@nalimilan I'm not against using something similar to the real Julia syntax for documentation but I think we need to stick very close to the actual syntax to avoid confusion for new Julians:
```julia
splitdir(path::AbstractString) :: (AbstractString, AbstractString)  # Bad, invalid syntax
splitdir(path::AbstractString) :: Tuple{AbstractString, AbstractString}  # Good

mkdir(path[, mode])  # Bad
mkdir(path, mode=0o777)  # Good, actual Julia syntax
```

**Comment by felipenoris** [comment]
I'm with @omus. Also, I propose we use actual Julia syntax, like

splitdir(path::AbstractString)::Tuple{AbstractString, AbstractString}

This will open space for automation.

**Comment by felipenoris** [comment]
I see that I may be stepping in some bigger discussion. The original idea of this PR was simply substitute `->` for `→` and follow along with the current documentation framework. Today we see `→` only on html generated files, but not on REPL. With this PR, `→` would be seen in REPL as well.

The idea I got with the current docs was that we should document "interfaces". Something like `mkdir(path[, mode])` may be due to `mkdir(path, mode=0o777)` method declaration, as well as method overload `mkdir(a)` , `mkdir(a, b)`.

So multiple dispatch really creates challenges for documentation.

So for this PR, I'll leave it open as is. If you think it's worth, go for it. If not, just close the PR.

File: doc/stdlib/collections.rst [hunk 1]
```diff
@@ -543,21 +543,21 @@ Iterable Collections
...
+.. function:: map(f, c...) → collection
 
    .. Docstring generated from Julia source
 
    Transform collection ``c`` by applying ``f`` to each element. For multiple collection arguments, apply ``f`` elementwise.
 
    .. doctest::
 
-       julia> map((x) -> x * 2, [1, 2, 3])
+       julia> map((x) → x * 2, [1, 2, 3])
```
Comment (omus) [inline doc/stdlib/collections.rst]: Anonymous functions can't be constructed with the unicode arrow

File: base/show.jl [hunk 2]
```diff
@@ -15,20 +15,20 @@ end
 """
     IOContext{<:IO} <: IO
 
-IOContext provides a mechanism for passing output-configuration keyword arguments through arbitrary show methods.
+`IOContext` provides a mechanism for passing output-configuration keyword arguments through arbitrary show methods.
 
-In short, it is an immutable Dictionary that is a subclass of IO.
+In short, it is an `immutable Dictionary` that is a subclass of `IO`.
```
Comment (omus) [inline base/show.jl]: I don't think "immutable Dictionary" should be in backticks

File: base/docs/helpdb/Base.jl [hunk 3]
```diff
@@ -9223,13 +9223,13 @@ that spans all rows of `R` in its corresponding QR factorization.
...
 """
-    map(f, c...) -> collection
+    map(f, c...) → collection
 
 Transform collection `c` by applying `f` to each element. For multiple collection arguments,
 apply `f` elementwise.
 
 ```jldoctest
-julia> map((x) -> x * 2, [1, 2, 3])
+julia> map((x) → x * 2, [1, 2, 3])
```
Comment (omus) [inline base/docs/helpdb/Base.jl]: Another anonymous function to fix

**Comment by omus** [comment]
@felipenoris sorry I got kind of side tracked with the conversation. The PR overall looks reasonable although I am a little worried how small the unicode right arrow is in the REPL in comparison to the ASCII arrow.

File: doc/stdlib/collections.rst (hunk 1 above)
Comment (felipenoris) [inline doc/stdlib/collections.rst]: Thanks! it's my mistake. I corrected most of these cases before commiting, but I let this one pass.

**Comment by tkelman** [comment]
There is ⟶ `\longrightarrow` but that won't show properly in the Windows console without manually changing to a non-default font.

**Comment by MichaelHatherly** [comment]
> little worried how small the unicode right arrow is in the REPL

Yes, it's a bit small maybe. How about using a long ascii arrow `-->` in the REPL and then in generated docs use Tony's `\longrightarrow` suggestion? Using `-->` instead of `->` would also help avoid syntax clashes with anonymous functions.

So we'd have the following four variations where `::` describes the return types and `-->` describes the names of the returned values:

```
func(x, y)
func(x, y) :: {Foo, Bar}
func(x, y) --> (foo, bar)
func(x, y) :: {Foo, Bar} --> (foo, bar)
```

This, of course, is assuming we get the `Tuple{...}` to `{...}` change proposed in #8470. `Tuple{...}` is pretty short anyway, so I'm not too bothered get that or not.

That said, these signatures should really just be a summary rather than being an exact 1-1 copy, and details that don't fit easily should be listed in the body of the docstring when necessary.

A lot of this can be automated away in time since we're storing all the needed information and have the mechanism in place to insert context-dependant generated content into docstrings already.

**Pull request closed without merging**